[EASI-3058] Calculate intake requester status

DylanSpOddball opened the pull request:
# EASI-3058

## Changes and Description

- Add a `SystemIntakeStatusRequester` enum to the GraphQL schema and a `statusRequester` field on `SystemIntake` of that type.
- Implement `CalculateSystemIntakeRequesterStatus()` to calculate that field based on an intake's state.
- Rename `NO_GOVERNANCE` to `NOT_GOVERNANCE` in IT Gov v2 enums, both in Go and the GraphQL schema.

## Notes
- I made extensive use of `switch` statements over possible enum values for step states, so I could benefit from [the exhaustiveness check I added](https://github.com/CMSgov/easi-app/pull/2104); I needed to return something in the `default` case, though, so most of my private functions return `(models.SystemIntakeStatusRequester, error)`.
- I made the enum name change from `NO_GOVERNANCE`  to `NOT_GOVERNANCE` to more accurately reflect what it means - that an intake was closed because it's not a governance request. It's a little inconsistent with the existing code, but I think the clarity is worth it; I'm open to talking about this, though.
  -  I only made the change for IT Gov v2-related enums, it shouldn't break any existing code. We'll need to let the team know, though.
- ~~Currently, an intake with `Step == DECISION_AND_NEXT_STEPS`, `DecisionState == NO_DECISION`, and `State == CLOSED` is treated as valid; calculating the status doesn't return an error. I followed [Steven's work](https://github.com/CMSgov/easi-app/pull/2096) on EASI-3048 on this, but it doesn't seem like a valid state, and maybe the status calculation should return an error? I think the only way that state should be reached is if an intake that was closed with no decision gets reopened, but once it hit the DECISION step, a proper decision should have been made, so `DecisionState` would be different. Thoughts?~~ **I updated this to make `Step == DECISION_AND_NEXT_STEPS`, `DecisionState == NO_DECISION` an invalid state regardless of whether the intake is open/closed.**

## How to test this change

1, Bring up the app backend.
2. Create a system intake:
```gql
mutation createSystemIntake {
    createSystemIntake(
        input: {requestType: MAJOR_CHANGES, requester: {name: "Mr. Fun Person"}}
    ) {
        id
    }
}
```
3. Query the created system intake:
```gql
query systemIntake {
    systemIntake(id: "systemIntakeIDFromAbove") {
        statusRequester
    }
}
```
It should return `"statusRequester": "INITIAL_REQUEST_FORM_NEW"`.

ClayBenson94 commented on pkg/graph/schema.graphql:
```diff
@@ -808,6 +808,7 @@ type BusinessCase {
 
 """
 The status of a system's IT governence request
+Note - pre-IT Gov v2 only - for IT Gov v2, use SystemIntakeRequesterStatus/SystemIntakeAdminStatus
```
Just some naming changes in this comment to match the type names
```suggestion
Note - pre-IT Gov v2 only - for IT Gov v2, use SystemIntakeStatusRequester/SystemIntakeStatusAdmin
```

DylanSpOddball replied:
Good catch; updated in https://github.com/CMSgov/easi-app/pull/2108/commits/74b63a81f42637496196fc07bff8a2b0a7efc102.

ClayBenson94 reviewed (changes requested):
These changes look great. I love the unit testing! One small comment change

ClayBenson94 commented:
@mynar7 Can you review the test cases and make sure everything looks good on your end? I think looking at these tests can probably be the majority of this review (and I didn't catch anything that looked off!)

mynar7 commented on pkg/graph/resolvers/system_intake_status_admin.go:
```diff
@@ -45,27 +45,27 @@ func calcSystemIntakeInitialFormStatusAdmin(intakeFormState models.SystemIntakeF
 
-func calcSystemIntakeDraftBusinessCaseStatusAdmin(draftBuisnessCaseState models.SystemIntakeFormState) models.SystemIntakeStatusAdmin {
-	if draftBuisnessCaseState == models.SIRFSSubmitted {
+func calcSystemIntakeDraftBusinessCaseStatusAdmin(draftBusinessCaseState models.SystemIntakeFormState) models.SystemIntakeStatusAdmin {
```
Good catch, thank you.

mynar7 commented on pkg/graph/resolvers/system_intake_status_admin.go:
```diff
@@ -45,27 +45,27 @@ func calcSystemIntakeInitialFormStatusAdmin(intakeFormState models.SystemIntakeF
 
-func calcSystemIntakeFinalBusinessCaseStatusAdmin(finalBuisnessCaseState models.SystemIntakeFormState) models.SystemIntakeStatusAdmin {
-	if finalBuisnessCaseState == models.SIRFSSubmitted {
+func calcSystemIntakeFinalBusinessCaseStatusAdmin(finalBusinessCaseState models.SystemIntakeFormState) models.SystemIntakeStatusAdmin {
```
We really need to get Steven to install that VS Code spellchecker haha

mynar7 commented on pkg/graph/resolvers/system_intake_status_admin.go:
```diff
@@ -87,8 +87,8 @@ func calcSystemIntakeDecisionStatusAdmin(decisionState models.SystemIntakeDecisi
 	}
-	if decisionState == models.SIDSNoGovernance {
-		return models.SISANoGovernance, nil
+	if decisionState == models.SIDSNotGovernance {
```
I agree that `NOT_GOVERNANCE` is more accurate and closer to the intent.

mynar7 commented on pkg/graph/resolvers/system_intake_status_requester.go:
```diff
@@ -0,0 +1,107 @@
+	case models.SystemIntakeStepDECISION:
+		return calcSystemIntakeDecisionStatusRequester(intake.DecisionState)
+	default:
+		return "", fmt.Errorf("issue calculating the requester intake status, no valid step")
```
Defaulting to an error is probably the right way to handle all these switch statements, nice job.

mynar7 commented on pkg/graph/resolvers/system_intake_status_requester.go:
```diff
@@ -0,0 +1,107 @@
+}
+
+func calcSystemIntakeGRTMeetingStatusRequester(grtDate *time.Time, currentTime time.Time) models.SystemIntakeStatusRequester {
+	if grtDate == nil || grtDate.After(currentTime) {
```
Shouldn't this status have a `READY_TO_SCHEDULE` for no date, `SCHEDULED` for a future date, and `AWAITING_DECISION` for a past date?

DylanSpOddball replied:
Not for the overall intake status - that only has "Ready for GRT meeting" and "Awaiting GRT decision" in the requester view. ([Figma](https://www.figma.com/file/ChzAP34A2DVvQUNQwD7lCt/IT-Governance-Next?type=design&node-id=475-115531&mode=design&t=foWugeQRbZN35Xtx-0)) The task list status for the GRT step will have the different values you're mentioning, that'll be covered in EASI-2960.

mynar7 replied:
That's weird, but OK, if that's what the designs call for. Thanks for clarifying.

mynar7 commented on pkg/graph/resolvers/system_intake_status_requester.go:
```diff
@@ -0,0 +1,107 @@
+
+func calcSystemIntakeFinalBusinessCaseStatusRequester(finalBusinessCaseState models.SystemIntakeFormState) (models.SystemIntakeStatusRequester, error) {
+	switch finalBusinessCaseState {
+	case models.SIRFSNotStarted, models.SIRFSInProgress:
```
Shouldn't the initial state return a `READY_TO_START`?

DylanSpOddball replied:
The overall intake status goes right to "Final Business Case in progress" ([Figma](https://www.figma.com/file/ChzAP34A2DVvQUNQwD7lCt/IT-Governance-Next?type=design&node-id=477-113294&mode=design&t=foWugeQRbZN35Xtx-0)). The task list status will have `READY_TO_START` as a possible value, that'll be in EASI-2961.

mynar7 replied:
That's weird, but OK, if that's what the designs call for. Thanks for clarifying.

mynar7 commented on pkg/graph/resolvers/system_intake_status_requester.go:
```diff
@@ -0,0 +1,107 @@
+}
+
+func calcSystemIntakeGRBMeetingStatusRequester(grbDate *time.Time, currentTime time.Time) models.SystemIntakeStatusRequester {
+	if grbDate == nil || grbDate.After(currentTime) {
```
Echoing above, shouldn't there be a `READY_TO_SCHEDULE` status for a `nil` date?

DylanSpOddball replied:
Not for overall intake status. ([Figma](https://www.figma.com/file/ChzAP34A2DVvQUNQwD7lCt/IT-Governance-Next?type=design&node-id=477-116633&mode=design&t=foWugeQRbZN35Xtx-0)) Task list status will be covered in EASI-2962.

mynar7 replied:
That's weird, but OK, if that's what the designs call for. Thanks for clarifying.

mynar7 commented on pkg/graph/resolvers/system_intake_status_requester.go:
```diff
@@ -0,0 +1,107 @@
+	case models.SIDSNotGovernance:
+		return models.SISRNotGovernance, nil
+	case models.SIDSNoDecision:
+		return "", fmt.Errorf("issue calculating the requester intake status - invalid state, intake decisionState is NO_DECISION, but intake is not closed")
```
I like this error message being super clear about what happened. I had suggested to Steven possibly putting something like this at the top of the file to make it very clear that we are checking for a specific invalid state, but I think this is a good way to handle it as well.

```
if (
  intake.Step ==  models.SystemIntakeStepDECISION &&
  intake.DecisionState == models.SIDSNoDecision
) return "", fmt.Errorf("invalid state...")
// switch
```

DylanSpOddball replied:
Yeah, at this point in the code we know _exactly_ what's invalid, so I figured we could have a pretty specific error message to help us debug if necessary.

mynar7 commented on src/types/graphql-global-types.ts:
```diff
@@ -237,6 +237,7 @@ export enum SystemIntakeState {
 
 /**
  * The status of a system's IT governence request
+ * Note - pre-IT Gov v2 only - for IT Gov v2, use SystemIntakeRequesterStatus/SystemIntakeAdminStatus
```
Good call

mynar7 reviewed (commented):
This all looks good outside of the questions I have about the meeting statuses.

DylanSpOddball commented:
FYI on testing - we added info to [our docs](https://github.com/CMSgov/easi-app/blob/main/docs/local_testing.md#server-tests) on how to run test functions that aren't part of a `testify` test suite; see the last bullet for "Server tests" at that link.

The unit tests follow the same pattern that Steven used in https://github.com/CMSgov/easi-app/pull/2096, I just reorganized the code a bit (and the test cases are different, since the logic's different).

mynar7 commented:
> FYI on testing - we added info to [our docs](https://github.com/CMSgov/easi-app/blob/main/docs/local_testing.md#server-tests) on how to run test functions that aren't part of a `testify` test suite; see the last bullet for "Server tests" at that link.
> 
> The unit tests follow the same pattern that Steven used in #2096, I just reorganized the code a bit (and the test cases are different, since the logic's different).

Dumb question, but if we're not using testify, these tests still run with the others in CI and locally just running the default test command, right?

DylanSpOddball commented:
> > FYI on testing - we added info to [our docs](https://github.com/CMSgov/easi-app/blob/main/docs/local_testing.md#server-tests) on how to run test functions that aren't part of a `testify` test suite; see the last bullet for "Server tests" at that link.
> > The unit tests follow the same pattern that Steven used in #2096, I just reorganized the code a bit (and the test cases are different, since the logic's different).
> 
> Dumb question, but if we're not using testify, these tests still run with the others in CI and locally just running the default test command, right?

Perfectly fine question! They're still run when running the whole test suite; `scripts/dev test:go` and CI both run all tests with `go test`, which picks up these test functions. The docs we added are just to clarify how to specify a _single_ test function to run when testing locally.

```sh
# This go test command is what scripts/dev test:go calls, plus the -v flag
$ go test -v -race -short -count=1 -p=1 ./... | grep "TestCalculateSystemIntakeRequesterStatus"

=== RUN   TestCalculateSystemIntakeRequesterStatus
=== RUN   TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Initial_Request_form_step
=== RUN   TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Initial_Request_form_step/Request_form_not_started
=== RUN   TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Initial_Request_form_step/Request_form_in_progress
=== RUN   TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Initial_Request_form_step/Request_form_edits_requested
=== RUN   TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Initial_Request_form_step/Request_form_submitted
=== RUN   TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Initial_Request_form_step/Closed_while_in_request_form
=== RUN   TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Draft_business_case_step
=== RUN   TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Draft_business_case_step/Draft_Biz_Case_form_not_started
=== RUN   TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Draft_business_case_step/Draft_Biz_Case_form_in_progress
=== RUN   TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Draft_business_case_step/Draft_Biz_Case_form_edits_requested
=== RUN   TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Draft_business_case_step/Draft_Biz_Case_form_submitted
=== RUN   TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Draft_business_case_step/Closed_while_in_Draft_Biz_Case_form
=== RUN   TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_GRT_meeting_step
=== RUN   TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_GRT_meeting_step/GRT_meeting_not_scheduled_yet
=== RUN   TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_GRT_meeting_step/GRT_meeting_scheduled_for_tomorrow
=== RUN   TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_GRT_meeting_step/GRT_meeting_happened_yesterday
=== RUN   TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_GRT_meeting_step/Closed_while_GRT_meeting_scheduled
=== RUN   TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Final_business_case_step
=== RUN   TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Final_business_case_step/Final_Biz_Case_form_not_started
=== RUN   TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Final_business_case_step/Final_Biz_Case_form_in_progress
=== RUN   TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Final_business_case_step/Final_Biz_Case_form_edits_requested
=== RUN   TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Final_business_case_step/Final_Biz_Case_form_submitted
=== RUN   TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Final_business_case_step/Closed_while_in_Final_Biz_Case_form
=== RUN   TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_GRB_meeting_step
=== RUN   TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_GRB_meeting_step/GRB_meeting_not_scheduled_yet
=== RUN   TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_GRB_meeting_step/GRB_meeting_scheduled_for_tomorrow
=== RUN   TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_GRB_meeting_step/GRB_meeting_happened_yesterday
=== RUN   TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_GRB_meeting_step/Closed_while_GRB_meeting_scheduled
=== RUN   TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Decision_step
=== RUN   TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Decision_step/Decision_made,_LCID_issued,_closed
=== RUN   TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Decision_step/Decision_made,_LCID_issued,_re-opened
=== RUN   TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Decision_step/Decision_made,_Not_Governance,_closed
=== RUN   TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Decision_step/Decision_made,_Not_Governance,_re-opened
=== RUN   TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Decision_step/Decision_made,_Not_Approved,_closed
=== RUN   TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Decision_step/Decision_made,_Not_Approved,_re-opened
=== RUN   TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Decision_step/Closed_with_no_decision_while_in_the_decision_step
=== RUN   TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Decision_step/Open_with_decisionState_set_to_No_Decision_while_in_the_decision_step
--- PASS: TestCalculateSystemIntakeRequesterStatus (0.01s)
    --- PASS: TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Initial_Request_form_step (0.00s)
        --- PASS: TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Initial_Request_form_step/Request_form_not_started (0.00s)
        --- PASS: TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Initial_Request_form_step/Request_form_in_progress (0.00s)
        --- PASS: TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Initial_Request_form_step/Request_form_edits_requested (0.00s)
        --- PASS: TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Initial_Request_form_step/Request_form_submitted (0.00s)
        --- PASS: TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Initial_Request_form_step/Closed_while_in_request_form (0.00s)
    --- PASS: TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Draft_business_case_step (0.00s)
        --- PASS: TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Draft_business_case_step/Draft_Biz_Case_form_not_started (0.00s)
        --- PASS: TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Draft_business_case_step/Draft_Biz_Case_form_in_progress (0.00s)
        --- PASS: TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Draft_business_case_step/Draft_Biz_Case_form_edits_requested (0.00s)
        --- PASS: TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Draft_business_case_step/Draft_Biz_Case_form_submitted (0.00s)
        --- PASS: TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Draft_business_case_step/Closed_while_in_Draft_Biz_Case_form (0.00s)
    --- PASS: TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_GRT_meeting_step (0.00s)
        --- PASS: TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_GRT_meeting_step/GRT_meeting_not_scheduled_yet (0.00s)
        --- PASS: TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_GRT_meeting_step/GRT_meeting_scheduled_for_tomorrow (0.00s)
        --- PASS: TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_GRT_meeting_step/GRT_meeting_happened_yesterday (0.00s)
        --- PASS: TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_GRT_meeting_step/Closed_while_GRT_meeting_scheduled (0.00s)
    --- PASS: TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Final_business_case_step (0.00s)
        --- PASS: TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Final_business_case_step/Final_Biz_Case_form_not_started (0.00s)
        --- PASS: TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Final_business_case_step/Final_Biz_Case_form_in_progress (0.00s)
        --- PASS: TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Final_business_case_step/Final_Biz_Case_form_edits_requested (0.00s)
        --- PASS: TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Final_business_case_step/Final_Biz_Case_form_submitted (0.00s)
        --- PASS: TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Final_business_case_step/Closed_while_in_Final_Biz_Case_form (0.00s)
    --- PASS: TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_GRB_meeting_step (0.00s)
        --- PASS: TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_GRB_meeting_step/GRB_meeting_not_scheduled_yet (0.00s)
        --- PASS: TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_GRB_meeting_step/GRB_meeting_scheduled_for_tomorrow (0.00s)
        --- PASS: TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_GRB_meeting_step/GRB_meeting_happened_yesterday (0.00s)
        --- PASS: TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_GRB_meeting_step/Closed_while_GRB_meeting_scheduled (0.00s)
    --- PASS: TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Decision_step (0.00s)
        --- PASS: TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Decision_step/Decision_made,_LCID_issued,_closed (0.00s)
        --- PASS: TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Decision_step/Decision_made,_LCID_issued,_re-opened (0.00s)
        --- PASS: TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Decision_step/Decision_made,_Not_Governance,_closed (0.00s)
        --- PASS: TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Decision_step/Decision_made,_Not_Governance,_re-opened (0.00s)
        --- PASS: TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Decision_step/Decision_made,_Not_Approved,_closed (0.00s)
        --- PASS: TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Decision_step/Decision_made,_Not_Approved,_re-opened (0.00s)
        --- PASS: TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Decision_step/Closed_with_no_decision_while_in_the_decision_step (0.00s)
        --- PASS: TestCalculateSystemIntakeRequesterStatus/Testing_statuses_for_the_Decision_step/Open_with_decisionState_set_to_No_Decision_while_in_the_decision_step (0.00s)
```

mynar7 commented:
> Perfectly fine question! They're still run when running the whole test suite; scripts/dev test:go and CI both run all tests with go test, which picks up these test functions. The docs we added are just to clarify how to specify a single test function to run when testing locally.

OK great, thank you for clarifying

mynar7 reviewed (dismissed):
This all looks good

DylanSpOddball commented:
After talking with @mynar7 a bit, I decided to have the status calculation function return an error if `step == DECISION` and `decisionState == NO_DECISION`, regardless of whether the intake is open or closed. This is an invalid state; if an intake has made it to the decision step, it should have had a proper decision made. I updated my code as well as the admin status calculation code to enforce this.

mynar7 reviewed (approved):
Nice, I like this way of handling invalid states.
